# Practica de Algebra lineal

### *Nombre*

In [1]:
import numpy as np

<span style="color:blue">Ejercicio 1:  
Usando los procedimientos add y mult_vector_escalar definidos anteriormente, calcular:   
{$\alpha$[1,2]+[3,4] : $\alpha$$\in$$\mathbb{R}$, 0$\leq$$\alpha$$\leq$1, con una precision de dos decimales (para la multiplicación)}</style>

In [2]:
def addn(v,w):
    return[v[i]+w[i] for i in range(len(v))]

In [3]:
def mult_vector_escalar(alpha, v):
    return [alpha*v[i] for i in range(len(v))]

In [4]:
def mult_y_suma(v, w):
    return [addn(mult_vector_escalar(alpha,v),w) for alpha in np.arange(0, 1.01, 0.01)]

In [5]:
v = [1,2]
w = [3,4]
mult_y_suma(v,w)

[[3.0, 4.0],
 [3.01, 4.02],
 [3.02, 4.04],
 [3.03, 4.06],
 [3.04, 4.08],
 [3.05, 4.1],
 [3.06, 4.12],
 [3.07, 4.14],
 [3.08, 4.16],
 [3.09, 4.18],
 [3.1, 4.2],
 [3.11, 4.22],
 [3.12, 4.24],
 [3.13, 4.26],
 [3.14, 4.28],
 [3.15, 4.3],
 [3.16, 4.32],
 [3.17, 4.34],
 [3.18, 4.36],
 [3.19, 4.38],
 [3.2, 4.4],
 [3.21, 4.42],
 [3.22, 4.44],
 [3.23, 4.46],
 [3.24, 4.48],
 [3.25, 4.5],
 [3.26, 4.52],
 [3.27, 4.54],
 [3.2800000000000002, 4.5600000000000005],
 [3.29, 4.58],
 [3.3, 4.6],
 [3.31, 4.62],
 [3.32, 4.64],
 [3.33, 4.66],
 [3.34, 4.68],
 [3.35, 4.7],
 [3.36, 4.72],
 [3.37, 4.74],
 [3.38, 4.76],
 [3.39, 4.78],
 [3.4, 4.8],
 [3.41, 4.82],
 [3.42, 4.84],
 [3.43, 4.86],
 [3.44, 4.88],
 [3.45, 4.9],
 [3.46, 4.92],
 [3.47, 4.94],
 [3.48, 4.96],
 [3.49, 4.98],
 [3.5, 5.0],
 [3.51, 5.02],
 [3.52, 5.04],
 [3.5300000000000002, 5.0600000000000005],
 [3.54, 5.08],
 [3.55, 5.1],
 [3.56, 5.12],
 [3.5700000000000003, 5.140000000000001],
 [3.58, 5.16],
 [3.59, 5.18],
 [3.6, 5.2],
 [3.61, 5.22],
 [3.62,

<span style="color:blue">Ejercicio 2:  
Mortadelo y Filemón usan como clave el siguiente vector:</style>  
> <span style="color:blue">**k**=[0,1,0,0,1,0,1,0,1,0]</style>  

<span style="color:blue">Mortadelo quiere enviarle a Filemón el siguiente mensaje:</style>  
> <span style="color:blue">**p**=[0,0,0,1,1,1,0,1,0,1]</style>  

<span style="color:blue">Mortadelo encripta su mensaje añadiendo k:</style>  
> <span style="color:blue">**c**=**p**+**k**=[0,0,0,1,1,1,0,1,0,1]+[0,1,0,0,1,0,1,0,1,0]=[0,1,0,1,0,1,1,1,1,1]</style>  

<span style="color:blue">Cuando Filemón recibe el mensaje, lo desencripta añadiendo **k** a lo que ha recibido</style>  
> <span style="color:blue">**p**=**c**+**k**=[0,1,0,1,0,1,1,1,1,1]+[0,1,0,0,1,0,1,0,1,0]=[0,0,0,1,1,1,0,1,0,1]</style>    

<span style="color:blue">que es el mensaje original</style>    
  
 <span style="color:blue">¿Cómo lo harías en Python? (Definiendo un procedimiento)</style>  

In [6]:
def suma_galois(u,v):
    return [(u[i]+v[i])%2 for i in range(len(u))]

In [7]:
#k es la clave que usan Mortadelo y Filemón
k=[0,1,0,0,1,0,1,0,1,0]
#p es el mensaje que Mortadelo le quiere enviar a Filemón
p=[0,0,0,1,1,1,0,1,0,1]

In [8]:
#c es el mensaje que recibe Filemón (Mortadelo ha encriptado el mensaje haciendo p+k)
c = suma_galois(p,k)
c

[0, 1, 0, 1, 0, 1, 1, 1, 1, 1]

In [9]:
#Cuando lo recibe, Filemón vuelve a añadir k para obtener el mensaje original p
p = suma_galois(c,k)
p

[0, 0, 0, 1, 1, 1, 0, 1, 0, 1]

<span style="color:blue">Ejercicio 3: coste total o beneficio. </style>

<span style="color:blue">Supongamos que D es el conjunto de algunos ingredientes de la cerveza: </style>
> <span style="color:blue">D={lúpulo, malta, agua, levadura}</style>  

<span style="color:blue">Por otro lado tenemos el vector coste:</style>  
> <span style="color:blue">coste={lúpulo: 2,5€, malta: 1.5€, agua: 0.006€, levadura: 0,45€}</style>  

<span style="color:blue">Por último tenemos el vector cantidad con lo necesario para hacer una cerveza:</style>  
> <span style="color:blue">cantidad={lúpulo: 6u, malta: 14u, agua: 7u, levadura: 11u}</style>  

<span style="color:blue">¿Cuánto cuesta hacer una cerveza?</style> 

In [10]:
D=['lupulo', 'malta', 'agua', 'levadura']
coste={'lupulo': 2.5, 'malta': 1.5, 'agua': 0.006, 'levadura': 0.45}
cantidad={'lupulo': 6, 'malta': 14, 'agua': 7, 'levadura': 11}

In [11]:
def coste_fabricar(D,u,v):
    return sum(u[d]*v[d] for d in D)

In [12]:
print('El coste de fabricar una cerveza es:', coste_fabricar(D,coste,cantidad), 'euros')

El coste de fabricar una cerveza es: 40.992000000000004 euros


<span style="color:blue"> Ejercicio 4:   
¿Cómo escribirías cada uno de los vectores {[3,0,0],[0,2,0],[0,0,1]} como combinación lineal de {[2,0,1],[1,0,2],[2,2,2]}?
Escribe las combinaciones lineales en Python. Seguramente tenga que definir una nueva suma.</style>

In [13]:
def suma3vectores(u,v,w):
    return[u[i]+v[i]+w[i] for i in range(len(v))]

In [14]:
#Como ya conozco el método solve, me ahorro calcular a mano los coeficientes ;-)
M = np.matrix([
    [2,1,2],
    [0,0,2],
    [1,2,2]
])
V = np.matrix([
    [3],
    [0],
    [0]
])
np.linalg.solve(M,V)

matrix([[ 2.],
        [-1.],
        [ 0.]])

In [15]:
#Esto mismo echo con el método suma3vectores sería:
u=[2,0,1]
v=[1,0,2]
w=[2,2,2]
suma3vectores(mult_vector_escalar(2,u),mult_vector_escalar(-1,v),mult_vector_escalar(0,w))

[3, 0, 0]

In [16]:
M = np.matrix([
    [2,1,2],
    [0,0,2],
    [1,2,2]
])
V = np.matrix([
    [0],
    [2],
    [0]
])
np.linalg.solve(M,V)

matrix([[-0.66666667],
        [-0.66666667],
        [ 1.        ]])

In [17]:
#Igual que antes, usando el método de suma3vectores
suma3vectores(mult_vector_escalar(-2/3,u),mult_vector_escalar(-2/3,v),mult_vector_escalar(1,w))

[0.0, 2.0, 0.0]

In [18]:
M = np.matrix([
    [2,1,2],
    [0,0,2],
    [1,2,2]
])
V = np.matrix([
    [0],
    [0],
    [1]
])
np.linalg.solve(M,V)

matrix([[-0.33333333],
        [ 0.66666667],
        [ 0.        ]])

In [19]:
#Una vez más, aplicando el método de suma3vectores
suma3vectores(mult_vector_escalar(-1/3,u),mult_vector_escalar(2/3,v),mult_vector_escalar(0,w))

[0.0, 0.0, 1.0]

<span style="color:blue"> Ejercicio 5:  
Dadas las matrices $
  A=
  \left[ {\begin{array}{cc}
   2 & -3 \\
   -4 & 6\\
  \end{array} } \right]
$ , $
  B=
  \left[ {\begin{array}{cc}
   8 & 4 \\
   5 & 5\\
  \end{array} } \right]
$
y $
  C=
  \left[ {\begin{array}{cc}
   5 & -2 \\
   3 & 1\\
  \end{array} } \right]
$, queremos verificar que AB=AC pero que B≠C.</style> 

In [20]:
A = np.matrix([
    [2,-3],
    [-4,6]
])
B = np.matrix([
    [8,4],
    [5,5]
])
C = np.matrix([
    [5,-2],
    [3,1]
])
A * B

matrix([[ 1, -7],
        [-2, 14]])

In [21]:
A * C

matrix([[ 1, -7],
        [-2, 14]])

El resultado es el mismo, sin embargo B ≠ C

<span style="color:blue"> Ejercicio 6 :  
Dadas la matriz $
  A=
  \left[ {\begin{array}{cc}
   3 & -6 \\
   -1 & 2\\
  \end{array} } \right]
$ , construye una matriz B de 2 × 2 tal que AB sea igual a la matriz cero. Las columnas de B no deben ser iguales entre sí y deben ser distintas de cero.
.</style> 

 $\left[ {\begin{array}{cc}
   3 & -6 \\
   -1 & 2\\
  \end{array} } \right]  
  *
  \left[ {\begin{array}{cc}
   x & y \\
   z & w\\
  \end{array} } \right]    
  =
    \left[ {\begin{array}{cc}
   0 & 0 \\
   0 & 0\\
  \end{array} } \right]  
$ 
Tengo que hayar la solución a las siguientes ecuaciones

3x-6z=0  
3y-6w=0  
-x+2z=0  
-y+2w=0  

Como podemos observar no es necesario ver que el rango de la matriz A es menor que dos para darnos cuenta de que las incógintas se relacionan entre sí de dos en dos (las x's con las z's y las y's con las w's)


In [22]:
#Luego el resultado será
#x=2z
#y=2w    
#Para que se cumpla la condición de que las columnas de B sean diferentes no podemos admitir entradas al procedimiento 
#de valores z=w y para que se cumpla la condición de que deben ser distintas de cero tampoco admito valores iguales a cero
def solucion_ejercicio_6(z,w):
    if z == 0 or w == 0:
        return 'No se admiten valores igual a cero'
    elif z == w:
        return 'No se admiten valores iguales para z y w'
    else:
        return (2*z,2*w)


Por ejemplo para una entrada z=1, w=2 tendríamos la solución (2,4):

In [23]:
solucion_ejercicio_6(1,2)

(2, 4)

Con lo que nos quedaría la siguiente matriz:  

$\left[ {\begin{array}{cc}
   2 & 4 \\
   1 & 2\\
  \end{array} } \right]  
$

<span style="color:blue"> Ejercicio 7:  
Sarrus para el determinante de orden 3.  
¿Cómo lo haríamos en Python?</style>

In [24]:
def sarrus(M):
    return ((M[0,0]*M[1,1]*M[2,2])+
            (M[0,1]*M[1,2]*M[2,0])+
            (M[0,2]*M[1,0]*M[2,1]))-((M[0,2]*M[1,1]*M[2,0])+
            (M[2,1]*M[1,2]*M[0,0])+
            (M[2,2]*M[1,0]*M[0,1]))
    

Comprobamos que funciona:

In [25]:
M = np.matrix([
    [2,1,2],
    [0,0,2],
    [1,2,2]
])
sarrus(M)

-6

In [26]:
np.linalg.det(M)

-6.0

<span style="color:blue"> Ejercicio 8:  
Sabemos que hay un modo directo en Python para calcular los rangos. Calcula el número de soluciones de estos tres sistemas de ecuaciones empleándolo.</style> 

> <span style="color:blue">x<sub>1</sub> − 2x<sub>2</sub> + x<sub>3</sub> = 0 </style>  
> <span style="color:blue">2x<sub>2</sub>−8x<sub>3</sub>= 8  </style>  
> <span style="color:blue">−4x<sub>1</sub> + 5x<sub>2</sub> + 9x<sub>3</sub> =− 9</style>

In [27]:
M = np.matrix([
    [1,-2,1],
    [0,2,-8],
    [-4,5,9]
])
MA = np.matrix([
    [1,-2,1,0],
    [0,2,-8,8],
    [-4,5,9,-9]
]) 
np.linalg.matrix_rank(M)

3

In [28]:
np.linalg.matrix_rank(MA)

3

Luego es un sistema compatible determinado con una única solución.

> <span style="color:blue">x<sub>2</sub> −4x<sub>3</sub> =8  </style>  
> <span style="color:blue">2x<sub>1</sub> −3x<sub>2</sub> + 2x<sub>3</sub> =1  </style>  
> <span style="color:blue">5x<sub>1</sub> −8x<sub>2</sub> + 7x<sub>3</sub> =1</style>

In [29]:
M = np.matrix([
    [0,1,-4],
    [2,-3,2],
    [5,-8,7]
])
MA = np.matrix([
    [0,1,-4,8],
    [2,-3,2,1],
    [5,-8,7,1]
]) 
np.linalg.matrix_rank(M)

2

In [30]:
np.linalg.matrix_rank(MA)

3

Luego es un sistema incompatible y no tiene solución

> <span style="color:blue">x<sub>1</sub> +x<sub>3</sub> =2  </style>  
> <span style="color:blue">2x<sub>1</sub> +3x<sub>2</sub> + x<sub>3</sub> =3  </style>  
> <span style="color:blue">3x<sub>1</sub> +3x<sub>2</sub> + 2x<sub>3</sub> =5</style>

In [31]:
M = np.matrix([
    [1,0,1],
    [2,3,1],
    [3,3,2]
])
MA = np.matrix([
    [1,0,1,2],
    [2,3,1,3],
    [3,3,2,5]    
]) 
np.linalg.matrix_rank(M)

2

In [32]:
np.linalg.matrix_rank(MA)

2

Luego es un sistema compatible indeterminado, es decir, tiene múltiples soluciones.

<span style="color:blue"> Ejercicio 9:  
Comprobar que r(x,y) para un ángulo $\theta$ es una función lineal.</style> 

Para ver que se trata de una función lineal debemos comprobar si estas dos propiedades se cumplen:
* Para cada vector v en V y cada escalar α en F, f (αv) = αf(v)  
* Por cada dos vectores u y v en V, f (u + v) = f (u) + f (v)  

r<sub>$\theta$</sub>($\alpha$x,$\alpha$y) = $
  \left[ {\begin{array}{cc}
   cos\theta & -sin\theta \\
   sin\theta & cos\theta \\
  \end{array} } \right] 
  *
  \left[ {\begin{array}{cc}
   \alpha x \\
   \alpha y \\
  \end{array} } \right]
  =
  \left[ {\begin{array}{cc}
   αx(cos\theta)-αy(sin\theta) \\
   αx(sin\theta)+αy(cos\theta) \\
  \end{array} } \right]
  =
  α *  \left[ {\begin{array}{cc}
   x(cos\theta)-y(sin\theta) \\
   x(sin\theta)+y(cos\theta) \\
  \end{array} } \right]
  = $
  $\alpha$ * r<sub>$\theta$</sub>(x,y)


r<sub>$\theta$</sub>([(x<sub>1</sub>,y<sub>1</sub>)+(x<sub>2</sub>,y<sub>2</sub>)]) =
$
  \left[ {\begin{array}{cc}
   cos\theta & -sin\theta \\
   sin\theta & cos\theta \\
  \end{array} } \right] 
  *
  \left[ {\begin{array}{cc}
   x_{1}+x_{2} \\
   y_{1}+y_{2} \\
  \end{array} } \right]
  =
  \left[ {\begin{array}{cc}
   (x_{1}+x_{2}) * cos\theta-(y_{1}+y_{2})*sin\theta \\
   (x_{1}+x_{2})*sin\theta+(y_{1}+y_{2})*cos\theta \\
  \end{array} } \right]
  \left[ {\begin{array}{cc}
   x_{1}*cos\theta+x_{2}*cos\theta-y_{1}*sin\theta-y_{2}*sin\theta \\
   x_{1}*sin\theta+x_{2}*sin\theta+y_{1}*cos\theta+y_{2}*cos\theta \\
  \end{array} } \right]
  =  
  \left[ {\begin{array}{cc}
   x_{1}*cos\theta-y_{1}*sin\theta+x_{2}*cos\theta-y_{2}*sin\theta \\
   x_{1}*sin\theta+y_{1}*cos\theta+x_{2}*sin\theta+y_{2}*cos\theta \\
  \end{array} } \right]=  
  \left[ {\begin{array}{cc}
   x_{1}*cos\theta-y_{1}*sin\theta \\
   x_{1}*sin\theta+y_{1}*cos\theta \\
  \end{array} } \right] +
  \left[ {\begin{array}{cc}
   x_{2}*cos\theta-y_{2}*sin\theta \\
   x_{2}*sin\theta+y_{2}*cos\theta \\
  \end{array} } \right] =  
  $  
  r<sub>$\theta$</sub>(x<sub>1</sub>,y<sub>1</sub>) + r<sub>$\theta$</sub>(x<sub>2</sub>,y<sub>2</sub>)

<span style="color:blue"> Ejercicio 10:  
¿Cuál es la representación coordenada del vector v= [6, 3, 2, 5] en términos de los vectores [2, 2, 2, 3], [1, 0, −1, 0], [0, 1, 0, 1]?</style> 

In [33]:
A=np.matrix([
    [2,1,0],
    [2,0,1],
    [2,-1,0],
    [3,0,1]
])
B=np.matrix([
    [2,1,0,6],
    [2,0,1,3],
    [2,-1,0,2],
    [3,0,1,5]
])
np.linalg.matrix_rank(A)
np.linalg.matrix_rank(B)
#El rango de ambas matrices es 3, luego es un SCD, una única solución, luego paso a resolverlo y obtengo 
#las coordenadas del vector [6,3,2,5] quitando una ecuación

3

In [34]:
A=np.matrix([
    [2,1,0],
    [2,0,1],
    [2,-1,0],    
])
B=np.matrix([
    [6],
    [3],
    [2]
])
np.linalg.solve(A,B)

matrix([[ 2.],
        [ 2.],
        [-1.]])

<span style="color:blue"> Ejercicio 11:  
Supongamos que queremos diseñar la red de internet para el otro campus universitario.  
La red debe lograr la misma conectividad que el grafo de entrada.  
Una arista representa un posible cable.  
El peso de la arista es el coste de instalar el cable.  
Nuestro objetivo es minimizar el coste total, usando el algoritmo Grow y el algoritmo Shrink.  </style> 

**Algoritmo de Grow:**

Ordenamos las aristas en orden creciente según sus pesos: {Bio-Med,Pembroke-Campus}, {Pembroke-Campus,Athletic-Complex}, {Athletic-Complex,Bio-Med}.

Partimos del conjunto vacío de nodos. Tomamos la primera arista, como todavía estos nodos no están conectados, la agregamos a nuestra solución. 

S = ({Bio-Med, Pembroke-Campus})

Para llegar a Ahtletic-Complex, se puede hacer a través de

* {Pembroke-Campus,Athletic-Complex}
* {Bio-Med,Athletic-Complex}

La que menos pesa es la primera, por lo que añado esta y deshecho la última arista:

S=({Bio-Med, Pembroke-Campus},{Pembroke-Campus,Athletic-Complex})




**Algoritmo de Shrink**

Ordenamos las aristas en orden decreciente según sus pesos: {Athletic-Complex,Bio-Med}, {Pembroke-Campus,Athletic-Complex},{Bio-Med,Pembroke-Campus}.

Partimos del conjunto de todas las aristas. 

S=({Athletic-Complex,Bio-Med}, {Athletic-Complex,Pembroke-Campus},{Pembroke-Campus,Bio-Med})

En la primera iteración vemos que podemos descartar la arista que va de Athletic-Complex a Bio-Med ya que existe un camino para llegar a Bio-Med. Luego,

S=({Athletic-Complex,Pembroke-Campus},{Pembroke-Campus,Bio-Med})

Y ya no puedo quitar más aristas, por lo que S es la solución.


<span style="color:blue"> Ejercicio 12:  
Sea la matriz $
  M=
  \left[ {\begin{array}{cc}
   1 & 0  & 0 & 5 \\
   0 & 2  & 0 & 7 \\
   0 & 0  & 3 & 9 \\
  \end{array} } \right]
$. Calcular el rango por filas y por columnas usando Python.</style> 

In [35]:
M = np.matrix([
    [1, 0, 0, 5],
    [0, 2, 0, 7],
    [0, 0, 3, 9]
])
rangoM = np.linalg.matrix_rank(M)
rangoM

3

In [36]:
MT = np.transpose(M)
rangoMT = np.linalg.matrix_rank(MT)
rangoMT

3

<span style="color:blue"> Ejercicio 13 (opcional):  
Crea un algoritmo que transforme una matriz cualquiera en GF(2) en una matriz escalonada.</style> 

Debemos buscar para cada columna las filas que no tienen un cero en esa columna y si hay más de una, coger la primera y transformar el resto para que tengan ceros en esa posición.

In [37]:
def forma_escalonada(M):
    cols = set(range(len(M[0])))
    MEscalonada = []
    filas_restantes = set(range(len(M)))
    for c in cols:
        filas_sin_ceros_en_c = [r for r in filas_restantes if M[r][c] != 0]
        if filas_sin_ceros_en_c != []:
            pivot = filas_sin_ceros_en_c[0]
            MEscalonada.append(M[pivot])
            filas_restantes.remove(pivot)
            for r in filas_sin_ceros_en_c[1:]:
                #Como estoy en GF(2) la unica posibilidad es que las dos filas tengan un 1 en la columna c, luego las 
                #sumo para obtener un cero
                M[r] = suma_galois(M[r],M[pivot])
    return MEscalonada        

Lo probamos con la siguiente matriz:

In [38]:
M=[[1,0,1],[0,1,1],[1,1,1]]

In [39]:
forma_escalonada(M)

[[1, 0, 1], [0, 1, 1], [0, 0, 1]]

Este procedimiento devuelve la matriz escalonada sin las filas que se quedan todas a cero, por ejemplo con la siguiente matriz obtenemos una escalonada con una fila menos, que sería toda de ceros:

In [40]:
M=[[1,0,1],[0,1,1],[1,1,0]]

In [41]:
forma_escalonada(M)

[[1, 0, 1], [0, 1, 1]]

En este caso sabemos que estas dos filas forman una base para el espacio filas.

<span style="color:blue"> Ejercicio 14:  
Modificar lo necesario en los algoritmos para que funcionen y calculen el problema de la casa en llamas.</style>

Definimos el producto interno en R<sup>2</sup> para los vectores u=(u<sub>1</sub>,u<sub>2</sub>) y v=(v<sub>1</sub>,v<sub>2</sub>) como el escalar: u<sub>1</sub> * v<sub>1</sub> + u<sub>2</sub> * v<sub>2</sub>

In [42]:
def producto_interno(u,v):
    return (u[0]*v[0] + u[1]*v[1])

Por lo tanto, la fórmula de la proyección paralela que vimos en clase quedaría así:

In [43]:
def proyeccion_paralela(b,a):
    sigma = (producto_interno(b,a))/(producto_interno(a,a)) if producto_interno(a,a) > 10^-20 else 0
    return mult_vector_escalar(sigma, a)

Aplicando esto al caso concreto de la casa en llamas, tenemos el siguiente resultado:

In [44]:
a=[6,2] #Es la calle
b=[2,4] #Es el lugar donde está la casa
proyeccion_paralela(b,a)

[3.0, 1.0]

Para la proyección ortogonal, debemos definir previamente la resta de vectores:

In [45]:
def resta_vectores(v,w):
    return [v[i]-w[i] for i in range(len(v))]

In [46]:
def proyeccion_ortogonal(b,a):
    return resta_vectores(b,proyeccion_paralela(b,a))

Y en el caso de la casa en llamas tenemos el siguiente resultado:

In [47]:
proyeccion_ortogonal(b,a)

[-1.0, 3.0]

Para averiguar si la distancia es menor de 3.5 y por tanto si la manguera llegaría a la casa, debemos obtener la norma de la proyección ortogonal. Sabemos que la raíz cuadrada de un número n es n<sup>1/2</sup>, con lo que me quedaría la siguiente definición de la norma:

In [48]:
def norma(v):
    return (v[0]**2+v[1]**2)**(1/2)

In [49]:
norma(proyeccion_ortogonal(b,a))

3.1622776601683795

Vemos que la distancia es 3.1622776601683795, que es menor que 3.5, luego la casa se salva.

**Feedback**: Para mi las clases han estado bien, las he podido seguir, alguna cosa que me quedó un poco menos clara la he asimilado al volver a ver los vídeos. Personalmente prefiero lo que se hizo en las últimas clases, para mi lo ideal es descansar unos diez minutos entre las dos sesiones e intentar optimizar al máximo el tiempo, si se puede terminar un poco antes mejor (cuatro horas de clase de mates de 20:30 a 00:30 es muy heavy :-)), siempre y cuando se vea todo el temario.